storage/badger/epoch_setups.go

```diff
@@ -38,13 +46,21 @@ func NewEpochSetups(collector module.CacheMetrics, db *badger.DB) *EpochSetups {
 			withStore(store),
 			withRetrieve(retrieve),
 			withResource(metrics.ResourceEpochSetup)),
+		lookup: make(map[uint64]*viewRange),
 	}
 
 	return es
 }
 
 func (es *EpochSetups) StoreTx(setup *flow.EpochSetup) func(tx *badger.Txn) error {
-	return es.cache.Put(setup.ID(), setup)
+	return func(tx *badger.Txn) error {
+		err := es.cache.Put(setup.ID(), setup)(tx)
+		if err != nil {
+			return err
+		}
+		es.lookup[setup.Counter] = &viewRange{first: setup.FirstView, last: setup.FinalView}
+		return nil
+	}
```

Comment:
⚠️  I'm concerned about the crash-resilience of this approach. 
```
           `EpochSetup` inserted here
           |
           v
| Staking | Setup | Committed |
```

`EpochSetup`s are inserted to the database shortly after they are emitted, effectively at the beginning of the epoch setup phase. This in-memory `lookup` will be populated at that point as well. If we restart at any time after that, `lookup` will not contain the view values any more. 

Instead, we could use an approach where we lazily retrieve start/end view for epochs when they are requested. There is actually a very similar procedure needed for handling the leader selection changes across epochs which is implemented [here](https://github.com/onflow/flow-go/blob/6c0e589ea2803b15db40adf09d11a975c5ff7d4a/consensus/hotstuff/committees/consensus_committee.go#L26).

Reply:
Good point.
In my latest commit, I opted for loading the in-memory `lookup` upon instantiating the `EpochSetups` storage. So if the node crashes and comes back up, it will repopulate the lookup cache with items from the persistant db.
This loading operation is performed in the constructor, as it feels unsafe to separate them. The caveat is that I currently ignore any errors potentially encountered during the loading process.. they will surface eventually when we fail to find the epoch corresponding to a view that should have a known epoch. An other option would be to return an error from the constructor, but that seems to go against a convention in Storage objects:
https://github.com/onflow/flow-go/blob/4688bbdea7bbe61558647ad376e24161ea5cf36d/cmd/scaffold.go#L382-L395
Any thoughts @jordanschalm ?

Reply:
Thanks for making the update. Unfortunately I think we have a separate issue with the new approach.

`EpochSetup`s are inserted to the database when the seal (soon, this will be receipt) referencing the block in which they were emitted is inserted to the database. `EpochSetup` events are fork-dependent. Each fork which straddles the final view of the auction will contain a separate `EpochSetup` event. Also, it is possible for malicious execution receipts containing an `EpochSetup` event to be included in a block and inserted to the database. 

This means in general we may have multiple "proposed" `EpochSetup` events for a given epoch in the database, but only one will be considered canonical once the requisite block has been sealed. Because of this, scanning the key prefix for `EpochSetup` events can result in reading `EpochSetup` events which are either malicious or non-canonical.

```
B1 <- B2 <- | <-B3a
              `-B3b
```
Both `B3a` and `B3b` could result in different execution results which each contain different EpochSetup events. Also, there may be both a valid result for (eg.) `B3a` as well as a malicious result, each with separate EpochSetup events. All these `EpochSetup` will be inserted to storage.

Reply:
The `protocol.State` API manages all this complexity for us, so I feel it would be better to use that as dependency for this component. 

The `protocol.State#Epochs` object exposes previous, current, and next epochs, which should be all we need. In general we can't guarantee that a node will have access to epoch data beyond these three, so it is safe to throw an error for a query that doesn't within the view bounds of these three epochs (even if the node _does_ happen to have that stored in the underlying storage) -- these queries indicate a bug in the querier.

For example, this would work:
```
func EpochForView(view uint64) (epochCounter uint64, err error) {
  current := state.Epochs().Current()
  next := state.Epochs().Next()
  previous := state.Epochs().Previous()

  for _, epoch := range []protocol.Epoch{previous, current, next} {
    if view >= epoch.FinalView() && view <= epoch.FirstView() {
      return epoch.Counter()
    }
  }

  return 0, errInvalidInput
}
```

Reply:
Good point. I reversed the changes to EpochSetups and used State directly in the SignerStore